src/Test.UnitTests.Sarif.Converters/TestData/Hdf/ExpectedOutputs/ValidResults.sarif

```diff
@@ -1204,6 +1326,7 @@
         },
         {
           "ruleId": "30002",
+          "level": "warning",
```

Comment:
>warning [](http://example.com/codeflow?start=20&length=7)

Your change in the main looks promising, but this outcome is clearly wrong, isn't it? This rule has no configuration, so the default value is warning already, and yet you've made that level explicit here. It shouldn't exist at all, isn't that correct?

Let's get on a phone call to discuss. I have an idea for you on how to test this. What you need here are very specific SARIF tests that comprehensively cover the spec. An ideal way to go about this is to author SARIF JSON and encode the expected failure level in the SARIF (via some convention in result production or as part of the property bag). Then you simply load the SARIF, verify that the API agrees the effective failure level is correct. Then serialize and deserialize the SARIF log and run the same check (to ensure that round-tripping works).

As a next step, let's author this test content and get it wired into a unit-test. This is the test-driven approach. Once we agree we've got all cases covered, well, when that test passes, you are done.

Reply:
This rule has no configuration, so the default value is warning already, and yet you've made that level explicit here. It shouldn't exist at all, isn't that correct?
----- my understanding was user want to able able to tell if user explicitly set the level, or it is not set and took the default value.
[https://github.com/microsoft/sarif-sdk/pull/2383](https://github.com/microsoft/sarif-sdk/pull/2383)

we made the Level field internally nullable to track if the field is explicitly set by user, if set we will write to the Sarif. if user not set it, it will not be written and the reader of the Sarif file should interpret the effective level base on the Sarif spec Logic.

The problem you point out is what happen if user  explicitly set the Level, and it is the same as default configuration(exist or not)? I think if we don't write it, it will be the same problem user can not tell so I did not check to add any logic for that when we write to Sarif file. Let me know,

Let's get on a phone call to discuss --- yes!

let's author this test content and get it wired into a unit-test. ---- agree and we do need it to have round trip